Automattic/woocommerce-payments
# Sending payment_method_update_data param to intent API

Fixes #5951
Fixes #8158

#### Changes proposed in this Pull Request

- Replaced `get_billing_details_from_order()` usage with `$order_service->get_billing_data_from_order()` so we can use `WC()` as it is not in Utils.
- Updated `$order_service->get_billing_data_from_order()` logic similar to how it was done before 7.5.2, but uses the Country method instead of the Checkout method to get the available fields.
- Sets `get_fees` return type to `array` to fix an error in tests `array_keys() expects parameter 1 to be array` in `includes/class-wc-payment-gateway-wcpay.php:3946`.
- Sets `get_account_default_currency` return type to `string` to fix a similar error.

#### Testing instructions

##### Sending payment_method_update_data param to intent API

1. Make sure your server is running the branch from the PR above
1. Place an order saving a new card with a billing address
1. Place an order using the saved card from the previous step but using a new billing address
1. Check the intents in Stripe dashboard and ensure it used the new billing address in the second payment

##### Avoid updating billing details for legacy card objects

``` 
💡 To create the legacy card object for your account, send the following request while inserting your customer ID, account ID and test API key

curl https://api.stripe.com/v1/customers/${customer_id}/sources \
-u ${test_api_key}: \
-H "Stripe-Account:${acct_id}" \
```
* Ensure that you have the saved payment method on your store which is legacy card from the request above
* On this branch, purchase a subscription with this saved payment method, confirm the successful purchase 
* Revert the fix commit, navigate to subscription details, process renewal, confirm that the error message is the same as in https://github.com/Automattic/woocommerce-payments/issues/8405
* Remove the revert so that the fix is in place
* In the failed order details options, choose `Retry renewal payment`
* Navigate to subscription details, process renewal and confirm the renewal was successful this time

##### Remove filtering of billing address details

1. Make a purchase saving a card using a billing address with address line 2.
2. Make a new purchase using the saved card from before but with an empty line 2.
3. Check Stripe dashboard. The second address should not have line 2.

##### Filtering billing details from order based on checkout fields

1. Create a simple subscription product
3. Install [Checkout Field Editor](https://woo.com/products/woocommerce-checkout-field-editor/?aff=10486&cid=1131038)
4. Remove all billing fields except ZIP code, email, and first/last name
5. Purchase a subscription via shortcode checkout (CFE is not compatible with the checkout block)
7. Process a renewal by running the woocommerce_scheduled_subscription_payment hook in the action scheduler.
8. The renewal should work

##### Subscription renewals can fail due to "meta_exists() on null" action scheduler error

> ❗ Important

Follow the reproduction steps from #8678 using WooPayments 7.5.1 and ensure you can reproduce the error, then switch to this branch and ensure it doesn't fail.

-------------------


**Post merge**
